### ISSUE TITLE
Why *Set classes?

### ISSUE THREAD

**Issue by kjetilk** [solid/data-interoperability-panel#98]
I had a look at the App Interop Spec dated 2021-03-15. I can't say I read the whole thing attentively, it is pretty big :-) , but at least the first part, I got a fairly good impression of what you're doing and I like the general direction.

There is one thing that I didn't understand though, why do you introduce all these classes with "sets"? Like `interop:ApplicationRegistrySet`, `interop:DataRegistrySet`?

RDF semantics are sets by default, so it seems to me like superfluous complexity. 

I get it that you'd like to first enumerate the registries, and then their registrations, but if there is a requirement to keep data from different registries apart, wouldn't that be better done by distinguishing properties of the different registrations rather than using the enumeration to do that?

**Comment by justinwb** [comment]
The reason we have `RegistrySet` instead of just `Registry` is to minimize information disclosure from the public profile document. Excerpt [from the spec](https://solid.github.io/data-interoperability-panel/specification/#agents-datamodel-summary):

> An Agent links to each kind of Registry Set through type-specific subproperties of interop:hasRegistrySet.
>
> This allows one Agent to link to many Registries across the Web, without exposing those Registries in a public document. Different permissions can be assigned to each Registry Set, depending on the sensitivity of the Registries they link to.
>
> Registry Sets link to Registries through the interop:hasRegistry property.

If we linked to each registry directly from the profile document, that would mean that we have to publicly advertise where all of our registries are, across different pods. The RegistrySet acts as an intermediate type, stored in an intermediate resource, that allow us to limit disclosure and have more fine-grained permissions on who knows what. Using this approach we can have a public document that lets you follow your nose if you have permission, but otherwise to the public looks like this:

```turtle
    a interop:Agent ;
    interop:hasApplicationRegistrySet <./application#set> ;
    interop:hasDataRegistrySet <./data#set> ;
    interop:hasAccessGrantRegistrySet <./grant#set> ;
    interop:hasAccessReceiptRegistrySet <./receipt#set> ;
    interop:hasRemoteDataRegistrySet <./remote#set> .
```

**Comment by kjetilk** [comment]
Right, OK, I can see the case for that. I am a bit worried about the additional complexity introduced, that might give a performance hit when querying the data. Lets keep an eye open for that as we move along.

**Comment by elf-pavlik** [comment]
## TL;DR

* we should evaluate who we expect to *follow their nose* using those registry sets - Applications, Authorization Agent, anything else?
* we should clarify who bootstraps initial Data Registry in user's first Storage, who let's them create Data Registries as they add more instances of Storage to their setup, where do they manage which Data Registration goes in which Registry (suggestion - their Authorization Agent)

## Details

I believe we may want to rethink this approach and focus on who would follows their nose using it as entry point. It looks that regular applications follow their nose via Access Receipts made available to them in their Application Registration. In that case they only need a way to find their registration, I have dedicated issue for it #91 and in #97 I show special case where Application could get link to their Application Registration in `id_token`.

When it comes to Authorization Agent, it could possibly also have it's own Application Registration and Access Receipts for all the data. It would probably self-maintain and self-issue those Access Receipts (with Trusted Grants?).

Besides looking at who else would need to follow their nose, I think we should also consider which party has responsibility for bootstrapping the initial data registrations (including sets if still needed and registries). In my current thinking about TypeScript implementation I tend to assume that Authorization Agent would include that functionality. I think operational solid setup would need to come provisioned with Authorization Agent, otherwise user can't authorize any other Application, including any other Trusted Agent.

In cases when user adds new Storage to their setup and want to start creating Data Registries in it, I think Authorization Agent would also take that responsibility. Assuming that new Storage has just empty root which user can access with their WebID, we should clarify how Authorization Agent will be able to access it on behalf of the user. I think special designation in user's WebID, similar to `solid:oidcIssuer` could be used to grant Authorization Agent full power of the user.

**Comment by elf-pavlik** [comment]
This will be addressed in #156

**Comment by justinwb** [comment]
Resolved by #165
